### PR TITLE
fix: use localstorage hook

### PR DESCRIPTION
**Pull request by sarkaaa**

### CODE REVIEW

File: src/app/components/calendar/calendarComponent.tsx [hunk 1]
```diff
@@ -1,25 +1,34 @@
...
 
 type Value = ValuePiece | [ValuePiece, ValuePiece];
 
-const CalendarComponent = ({ setCalendarHandle }: { setCalendarHandle: any }) => {
+const CalendarComponent = ({
+  setCalendarHandle,
+}: {
+  setCalendarHandle: (e: any) => void;
+}) => {
   const handleWeekNumber = (wN: number, date: Date) => {
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarComponent.tsx:15]:
```suggestion
  const handleWeekNumber = (weekNumber: number, date: Date) => {
```

also I would appreciate some documentation comments. As someone who's never seen this code before,
a short explanation like

```tsx
  // Called when the user selects (clicks on) a week number in the calendar
  const handleWeekNumber = (weekNumber: number, date: Date) => {
```

would be very helpful. Even better with a short description above the `CalendarComponent`
component itself:

```tsx
/**
 * Renders an interactive calendar. The user may interact by
 * selecting a specific day, or a whole week, by clicking on it
 * in the calendar.
 */
const CalendarComponent = ({
```

with this, the comment right above `handleWeekNumber` perhaps wouldn't even be necessary. I'd also argue in favor of renaming the function to `handleWeekClick` or `selectWeekNumber`, or similar. I find the current name confusing, because "weekNumber" is a thing, not an action to be handled.

File: src/app/components/calendar/weeklyCalendar.tsx [hunk 2]
```diff
@@ -6,6 +6,12 @@ import { WEEKDAYS, MONTHS } from "../../data/constants";
 
 const CalDates = Object.values(WEEKDAYS);
 
+const CANVAS_VARIANTS: any = {
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/weeklyCalendar.tsx:9]: Why are we adding untyped code? 🤔

Also nitpick but wdyt about `CANVAS_BACKGROUNDS`?

File: src/app/components/calendar/calendarForm.tsx [hunk 3]
```diff
@@ -25,34 +26,27 @@ const CalendarForm = () => {
...
+  }
+
+  const [calendarSetup, setCalendar] = useLocalStorage('calendarData', calendarDefaultValues)
+  const [calendarDates, setCalendarDates] = useLocalStorage('calendarDatesData', calendarDatesDefault)
+
+  const [mount, setMount]= useState(false)
 
   // TMP
-  const handleInput = (e: any) => {
+  const handleInput = (e: { target: { [key: string]: any }}) => {
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarForm.tsx:42]:
```suggestion
  const handleInput = (e: React.MouseEvent) => {
```

Overall, I see plenty of `any` 😅

File: src/app/components/calendar/calendarComponent.tsx [hunk 4]
```diff
@@ -1,25 +1,34 @@
...
-    const dayNumber =  date.getDate()
+    const dayNumber = date.getDate();
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarComponent.tsx:16]: There are plenty of style changes, and a lot of ESLint warning and errors across the project 😅 Does you IDE report them to you as you type?

Also, it clutters the MR quite a lot. Ideally, I'd say this would be in its own MR, or at least commit

File: src/app/components/calendar/calendarComponent.tsx [hunk 5]
```diff
@@ -1,25 +1,34 @@
...
+    });
+    setCalendarHandle({target: {name: 'year', value: date.getFullYear()}});
+  };
 
-    setCalendarHandle({ target: { name: 'weekNumber', value: wN }})
-    setCalendarHandle({ target: { name: 'weekDayNumbers', value: Array.from({length: 7}, (_, i) => dayNumber + i) }})
-    setCalendarHandle({ target: { name: 'year', value: date.getFullYear() }})
-  }
-  
   const handleMonth = (date: Date | Value) => {
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarComponent.tsx:28]: Unrelated but Date is a Value so this typing is confusing. Also Value is a pretty generic name, which leaves me guessing.

```suggestion
  const handleMonth = (date: Value) => {
```

Edit: ah, so it's actually from the Calendar library. Since imports do not work because of the way the library is set up (authors mistake ts ts), this would also be a way to access the library's definition, but in this particular case, I'd say it's not worth the complexity:

```tsx
import { NonUndefined } from 'react-hook-form';

// Value is the type of the first parameter of the onChange function of the Calendar component.
type CalendarOnChange = NonUndefined<React.ComponentProps<typeof Calendar>['onChange']>;
type Value = Parameters<CalendarOnChange>[0];
```

Maybe I'd just go with

```tsx
  // The library's typing also includes non-dates, but this seems never to happen.
  const handleMonth = (date: unknown) => {
    if (!(date instanceof Date) return;
    
    ...
  }
```

Wdyt? Leaving the original Value type is also an option

File: src/app/components/calendar/calendarCell.tsx [hunk 6]
```diff
@@ -3,7 +3,7 @@ import { libreBaskerville, raleway } from "../../utils/fonts";
 type Props = {
   title: string;
   dayNumber: number;
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarCell.tsx:5]: Clarity-wise: is this `dayOfWeek`, `dayOfMonth` or something else? Hard to find out! 😅

File: src/app/components/calendar/calendarComponent.tsx [hunk 7]
```diff
@@ -28,25 +37,30 @@ const CalendarComponent = ({ setCalendarHandle }: { setCalendarHandle: any }) =>
...
     }
-  }
+  };
 
-  return <Calendar showNeighboringMonth onClickWeekNumber={handleWeekNumber} onClickMonth={handleMonth} onChange={handleMonth} locale="en-GB" showWeekNumbers />
+  return (
+    <Calendar
+      showNeighboringMonth
+      onClickWeekNumber={handleWeekNumber}
+      onClickMonth={handleMonth}
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarComponent.tsx:58]: Unrelated from this MR, but from user experience perspective, I'm not sure it's the best idea to change the displayed calendar (the one to be printed) on month click. If you click on the month selection again, you'll notice that the old month is still selected, and if you select it back, you'll see that the old day is still active, so the old day was active the whole time.

Perhaps let's only change the calendar to be printed when the active day changes (ie. delete ` onClickMonth` and keep only `onChange` and `onClickNumber`)?

Also, notice that clicking the arrows next to the month (see pic) also do not change the selected day, which I think is correct.

image

File: src/app/components/calendar/calendarForm.tsx [hunk 8]
```diff
@@ -25,34 +26,27 @@ const CalendarForm = () => {
...
-  const [calendarDates, setCalendarDates] = useState({
+  const calendarDatesDefault = {
     weekNumber: null,
     weekDayNumbers: [],
     month: new Date().getMonth(),
     year: new Date().getFullYear(),
-  });
+  }
+
+  const [calendarSetup, setCalendar] = useLocalStorage('calendarData', calendarDefaultValues)
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarForm.tsx:36]: It is customary for the state variable and the state setter to have names X and setX. Eg:

```suggestion
  const [calendarSettings, setCalendarSettings] = useLocalStorage('calendarData', calendarDefaultValues)
```

File: src/app/components/calendar/calendarForm.tsx [hunk 9]
```diff
@@ -25,34 +26,27 @@ const CalendarForm = () => {
...
+  const [calendarDates, setCalendarDates] = useLocalStorage('calendarDatesData', calendarDatesDefault)
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarForm.tsx:37]: Nitpick: wdyt about `calendarState` / `setCalendarState`?

File: src/app/components/calendar/calendarForm.tsx [hunk 10]
```diff
@@ -25,34 +26,27 @@ const CalendarForm = () => {
...
-      return { ...prevState, [fieldName]: fieldValue };
-    };
-    let values = newVals(calendarSetup);
-
-    localStorage.setItem("calendarData", JSON.stringify(values));
+    setCalendar({ ...calendarSetup, [fieldName]: fieldValue })
   };
 
-  const setCalendarHandle = (e: any) => {
+  const setCalendarHandle = (e: { target: { [key: string]: any }}) => {
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarForm.tsx:49]: Also unrelated to this MR, but I have several observations about this function.

0. What is a calendar handle? Would `setCalendarState` work better?

1. The function is typed as if it expected a user event (ie. an instance of the [Event interface](https://developer.mozilla.org/en-US/docs/Web/API/Event), having a target prop etc.), However, it does not. I would recommend giving it params propName and value instead, since the "target" prop is useless here. Wdyt?

```suggestion
  const setCalendarState = (propName, value) => {
```

Even better, let's make accept the entire state of the calendar:

```suggestion
  const setCalendarState = (state: CalendarState) => {
```

There is also an issue of representation of the calendar state. All the properties (weekNumber, weekDayNumbers, month and year) are computable from the currently selected date. I would consider representing the state of the calendar as a single Date object, and computing the year, week and so on where you need it.

For example, you could have a single helper function for computing the days of the week that contains a date.
I see several places where you're using `Array.from({length: 7}, (_, i) => ...`, ideally this would be unified:

```
const getDaysOfWeek = (aDayInTheWeek: Date) => {
  // Let's use `Array.from({length: 7}, (_, i) => ...` here and nowhere else, if possible.
};
```

This function would be useful even if you decide to keep the current representation of calendar state.

Does this make sense to you?

File: src/app/components/calendar/calendarForm.tsx [hunk 11]
```diff
@@ -25,34 +26,27 @@ const CalendarForm = () => {
...
     weekDayNumbers: [],
     month: new Date().getMonth(),
     year: new Date().getFullYear(),
-  });
+  }
+
+  const [calendarSetup, setCalendar] = useLocalStorage('calendarData', calendarDefaultValues)
+  const [calendarDates, setCalendarDates] = useLocalStorage('calendarDatesData', calendarDatesDefault)
+
+  const [mount, setMount]= useState(false)
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/calendarForm.tsx:39]: What is the purpose of the `mount` variable`?

File: src/app/components/calendar/weeklyCalendar.tsx [hunk 12]
```diff
@@ -51,15 +57,15 @@ const WeeklyCalendar = ({ calendar, calendarSetup }: { calendar: any, calendarSe
     calendar.weekNumber = getWeekNumber(new Date());
   }
 
-  if (calendar.weekDayNumbers.length === 0) {
+  if (calendar.weekDayNumbers?.length === 0) {
     calendar.weekDayNumbers = defaultWeekDayNumbers();
   }
 
   let weekDays: number[] = fillWeekdays(calendar || defaultWeekDayNumbers(), []);
```
Comment (jozef-mikusinec-kiwicom) [inline src/app/components/calendar/weeklyCalendar.tsx:64]: Whoa, this is not intended, is it?

```suggestion
  let weekDays: number[] = fillWeekdays(calendar, []) || defaultWeekDayNumbers();
```

Adding a type to `fillWeekdays` would help discover this issue.

Relatedly, why is it even allowed for `weekDayNumbers` not to be present?

**Comment by sarkaaa** [comment]
I'll merge it as localStorage is set okeyish and "code-refactoring" fixes will be covered in the next PR.

File: src/app/components/calendar/calendarComponent.tsx (hunk 5 above)
Comment (sarkaaa) [reply src/app/components/calendar/calendarComponent.tsx:28]: yeah, my choice of props selection `Date | Value` is based on the library which is used for calendar component. I personally like the 2nd option as it looks very clear and pretty straightforward. Thank you. ❤️

File: src/app/components/calendar/calendarComponent.tsx (hunk 7 above)
Comment (sarkaaa) [reply src/app/components/calendar/calendarComponent.tsx:58]: yeah, thank you for the feedback, that makes sense. :)

File: src/app/components/calendar/calendarForm.tsx (hunk 11 above)
Comment (sarkaaa) [reply src/app/components/calendar/calendarForm.tsx:39]: I've tried to resolve this issue, however, I still got a loader component and also warning message in the console. According to this article, I think there is no better option than to set something like "mounting" useState. 🤔  @jozef-mikusinec-kiwicom 

https://dev.to/gvegacl/nextjs-server-prop-did-not-match-client-prop-and-conditional-styleclass-rendering-2i4c
